accord-core/src/main/java/accord/local/durability/DurabilityQueue.java

```diff
@@ -73,325 +86,702 @@ public class DurabilityQueue
+                    retry(pending);
+                }
             }
-            else if (request == null)
+            finally
             {
-                // if request != null, the request will log the necessary information
-                logger.debug("{}: Successfully achieved durability for {}.", txnId, ranges);
+                synchronized (this)
+                {
+                    Invariants.require(pending.status != COMPLETING);
+                }
             }
         });
     }
 
-    void retry(SyncPoint<Range> syncPoint, DurabilityRequest request, int attempt)
+    private void done(Pending done)
     {
-        long retryDelay = node.agent().retryDurabilityDelay(node, attempt, MICROSECONDS);
-        Invariants.require(retryDelay > 0);
-        if (request != null) logger.info("{}: Retrying durability for {} requested by {} in {}s.", syncPoint.syncId, syncPoint.route.toRanges(), request.requestedBy, String.format("%.3f", retryDelay/1000_000.0));
-        else logger.debug("{}: Retrying durability for {} in {}s.", syncPoint.syncId, syncPoint.route.toRanges(), String.format("%.3f", retryDelay/1000_000.0));
-        node.scheduler().selfRecurring(() -> submit(syncPoint, request, attempt), retryDelay, MICROSECONDS);
+        Invariants.require(done.deferred == null);
+        unregister(done);
+        done.status = DONE;
+        adapter.done(done.request, done.syncPoint);
     }
 
-    void restart(SyncPoint<Range> syncPoint, DurabilityRequest request, int attempt)
+    private void abandon(Pending abandon, boolean requeueDeferred, boolean pruned)
     {
-        long retryDelay = node.agent().retryDurabilityDelay(node, attempt, MICROSECONDS);
-        Invariants.require(retryDelay > 0);
-        logger.debug("{}: Restarting durability for {} in {}s.", syncPoint.syncId, syncPoint.route.toRanges(), String.format("%.3f", retryDelay/1000_000.0));
-        node.scheduler().selfRecurring(() -> node.durability().shards().request(request, request.stillWaiting(syncPoint.route)), retryDelay, MICROSECONDS);
+        Invariants.require(requeueDeferred || abandon.deferred == null);
+        Invariants.require(abandon.parent == null);
+        if (abandon.isInHeap())
+            queued.remove(abandon);
+        abandon.status = ABANDONED;
+        if (requeueDeferred)
+            requeueDeferred(abandon);
+        unregister(abandon);
+        adapter.abandon(abandon.request, abandon.syncPoint, pruned);
+    }
+
+    private synchronized void completing(Pending pending, long startedAt)
+    {
+        if (pending.status != ACTIVE || pending.startAt != startedAt)
+            return;
+
+        pending.status = COMPLETING;
+        --activeCount;
+        requeueDeferred(pending);
+        enqueue(requeue.poll());
+    }
+
+    private void requeueDeferred(Pending parent)
+    {
+        Invariants.require(parent.status.compareTo(COMPLETING) >= 0);
+        List<Pending> deferreds = parent.deferred;
```

Comment:
I am a bit confused after reading this method, since we are setting items by index to this array; do we need to call `truncate` in the end to remove all pre-existing items from there?